observatory-dags/observatory/dags/telescopes/crossref_metadata.py

```diff
@@ -14,289 +14,210 @@
-    :return: the list of CrossrefMetadataRelease instances.
+        logging.info(f'Transform input folder: {self.extract_folder}, output folder: {self.transform_folder}')
+        finished = 0
+        # Transform each file in parallel
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            futures = []
+
+            # List files and sort so that they are processed in ascending order
+            input_file_paths = natsorted(self.extract_files)
+
+            # Create tasks for each file
+            for input_file in input_file_paths:
+                # The output file will be a json lines file, hence adding the 'l' to the file extension
+                output_file = os.path.join(self.transform_folder, os.path.basename(input_file) + 'l')
+                future = executor.submit(transform_file, input_file, output_file)
+                futures.append(future)
+
+            # Wait for completed tasks
+            for future in as_completed(futures):
+                future.result()
+                finished += 1
+                if finished % 1000 == 0:
+                    logging.info(f'Transformed {finished} files')
+
+
+class CrossrefMetadataTelescope(SnapshotTelescope):
     """
+    The Crossref Metadata Telescope
 
-    return ti.xcom_pull(key=CrossrefMetadataTelescope.RELEASES_TOPIC_NAME,
-                        task_ids=CrossrefMetadataTelescope.TASK_ID_CHECK_RELEASE, include_prior_dates=False)
-
-
-class CrossrefMetadataTelescope:
-    """ A container for holding the constants and static functions for the crossref metadata telescope. """
+    Saved to the BigQuery table: <project_id>.crossref.crossref_metadataYYYYMMDD
+    """
 
     DAG_ID = 'crossref_metadata'
-    DATASET_ID = 'crossref'
-    DESCRIPTION = 'The Crossref Metadata Plus dataset: ' \
-                  'https://www.crossref.org/services/metadata-retrieval/metadata-plus/'
-    RELEASES_TOPIC_NAME = "releases"
-    QUEUE = 'remote_queue'
-    MAX_PROCESSES = cpu_count()
-    MAX_CONNECTIONS = cpu_count()
-    MAX_RETRIES = 3
-
     TELESCOPE_URL = 'https://api.crossref.org/snapshots/monthly/{year}/{month:02d}/all.json.tar.gz'
-    # DEBUG_FILE_PATH = os.path.join(test_data_path(), 'telescopes', 'crossref_metadata.json.tar.gz')
-
-    TASK_ID_CHECK_DEPENDENCIES = "check_dependencies"
-    TASK_ID_CHECK_RELEASE = f"check_release"
-    TASK_ID_DOWNLOAD = f"download"
-    TASK_ID_UPLOAD_DOWNLOADED = 'upload_downloaded'
-    TASK_ID_EXTRACT = f"extract"
-    TASK_ID_TRANSFORM = f"transform_releases"
-    TASK_ID_UPLOAD_TRANSFORMED = 'upload_transformed'
-    TASK_ID_BQ_LOAD = f"bq_load"
-    TASK_ID_CLEANUP = f"cleanup"
 
-    @staticmethod
-    def check_dependencies(**kwargs):
-        """ Check that all variables exist that are required to run the DAG.
+    def __init__(self, dag_id: str = DAG_ID, start_date: datetime = datetime(2020, 6, 1),
+                 schedule_interval: str = '@monthly', dataset_id: str = 'crossref',
```

Comment:
The original start_date was set to `datetime(2020, 6, 7)`, so we need to keep it the same because Airflow doesn't let you change the start date once the DAG has already run.

The reason why it is set to the 7th day of the month is because of when Crossref Metadata snapshots are released: 
> "New snapshots are created each month, available by the 5th day, providing all records up to and including the previous month."

See here: https://www.crossref.org/documentation/metadata-plus/metadata-plus-snapshots/#00344

I'm not actually sure how well they keep to that schedule though, I think we should check.

If we set the `schedule_interval` to `monthly`, then that will affect the DOI workflow, which runs weekly and has a sensor that waits for the Crossref Metadata DAG to finish. I think that the sensor will wait indefinitely on weeks when the Crossref Metadata DAG does not run. 

Would you be able to find out how you can get the Crossref Metadata sensor in the DOI DAG to only wait on weeks when the Crossref Metadata DAG runs?

Reply:
> The original start_date was set to `datetime(2020, 6, 7)`, so we need to keep it the same because Airflow doesn't let you change the start date once the DAG has already run.
> 
> The reason why it is set to the 7th day of the month is because of when Crossref Metadata snapshots are released:
> 
> > "New snapshots are created each month, available by the 5th day, providing all records up to and including the previous month."
> 
> See here: https://www.crossref.org/documentation/metadata-plus/metadata-plus-snapshots/#00344
 
I see, I have changed this back :-)

> I'm not actually sure how well they keep to that schedule though. Could we add some logging to the check_release task that prints out all of the available releases?

I'll have a look into adding some logging to the check_release task. I think we will have to get the HEAD of each URL to check if it exists, maybe we could limit it to 2 months before and 2 months after? I don't think there is a single URL with an overview all available releases.

> I think that there is a bug in the existing code on the develop branch, because the execution date is being used to fetch the the Crossref Metadata release. Should it be execution date - 1 month? E.g. the 2021-05 release is uploaded on 2021-06-05, so if the execution date was 2021-06-01, then it would never fetch the 2021-05 release. I think that is a problem because I set the DAG to run weekly. That wouldn't be a problem for monthly though right? Because the execution date would be for the previous month?

Yeah I think you are right. This will be solved if we set it to '@monthly' and use the 'prev_execution_date' to create the URL. In that case for the execution_date 2021-06-01, it will be running on 2021-07-01 and it will check for the release of 2021-05. 
Alternatively we could set the schedule to '0 0 7 * *' (so it is after the 5th day) and use the 'execution_date' to create the URL. In that case for the execution_date of 2021-06-07, it will be running on 2021-07-07 and it will check for the release of 2021-06. So we would be a bit more up date.

> If we set the `schedule_interval` to `monthly`, then that will affect 1) the DOI workflow, which runs weekly and has a sensor that waits for the Crossref Metadata DAG to finish. I think that the sensor will wait indefinitely on weeks when the Crossref Metadata DAG does not run. 2) The worker VM is required,
> 
> Would you be able to find out how you can get the Crossref Metadata sensor in the DOI DAG to only wait on weeks when the Crossref Metadata DAG runs?

Yeah by default the ExternalTaskSensor expects the schedule_interval+start date to be in sync. We could change this by using the 'execution_date_fn' parameter of the ExternalTaskSensor. I have added a solution in the latest commit.
What do you think of this solution? I have tested it with this simple DAG:  
```python
from datetime import datetime

from airflow import DAG
from airflow.operators.dummy_operator import DummyOperator
from airflow.sensors.external_task_sensor import ExternalTaskSensor
from observatory.platform.telescopes.telescope import Telescope, Release
from observatory.platform.utils.telescope_utils import get_prev_execution_date


class FakeCrossrefTelescope(Telescope):
    DAG_ID = 'fake_crossref'
    SCHEDULE_INTERVAL = '@monthly'

    def __init__(self, dag_id: str = DAG_ID, start_date: datetime = datetime(2020, 6, 7),
                 schedule_interval: str = SCHEDULE_INTERVAL, catchup=True):
        """ Construct a Telescope instance.

        :param dag_id: the id of the DAG.
        :param start_date: the start date of the DAG.
        :param schedule_interval: the schedule interval of the DAG.
        """
        super().__init__(dag_id, start_date, schedule_interval, catchup=catchup)

        self.add_setup_task_chain([self.setup_task])
        self.add_task(self.task)

    def make_release(self, **kwargs) -> Release:
        """ Create a single release instance.

        :param kwargs: the context passed from the PythonOperator. See
        https://airflow.apache.org/docs/stable/macros-ref.html for more info.
        :return: A release instance
        """
        release_id = f"telescope_{datetime.now()}"
        release = Release(self.dag_id, release_id)
        return release

    def setup_task(self, **kwargs) -> bool:
        """ This is a custom task that is executed before a release is made. The return value is used for a
        shortcircuit operator"

        :param kwargs: the context passed from the PythonOperator. See
        https://airflow.apache.org/docs/stable/macros-ref.html for more info.
        :return: True/False
        """
        setup = 'successful'
        return True if setup == 'successful' else False

    def task(self, release: Release, **kwargs):
        """ This is a custom task that can use a release instance.

        :param release: A release instance.
        :param kwargs: the context passed from the PythonOperator. See
        https://airflow.apache.org/docs/stable/macros-ref.html for more info.
        :return: None.
        """
        print(f"This is a custom task that uses the release: {release.release_id}")


telescope = FakeCrossrefTelescope()
globals()[telescope.dag_id] = telescope.make_dag()


default_args = {
    "owner": "airflow",
    "start_date": datetime(2020, 8, 30)
}
SENSOR_ID_CROSSREF_METADATA = 'crossref_metadata_sensor'

with DAG(dag_id='test_sensor', schedule_interval='@weekly', default_args=default_args, catchup=False) as dag2:
    # Sensors
    crossref_metadata_sensor = ExternalTaskSensor(
        task_id=SENSOR_ID_CROSSREF_METADATA,
        external_dag_id=FakeCrossrefTelescope.DAG_ID,
        execution_date_fn=lambda execution_date: get_prev_execution_date(FakeCrossrefTelescope.SCHEDULE_INTERVAL,
                                                                         execution_date),
        mode='reschedule')

    dummy = DummyOperator(
        task_id='dummy'
    )

    crossref_metadata_sensor >> dummy
```

What would be the issue for the worker VM if the schedule interval is not set to '@weekly'?

Reply:
> > The original start_date was set to `datetime(2020, 6, 7)`, so we need to keep it the same because Airflow doesn't let you change the start date once the DAG has already run.
> > The reason why it is set to the 7th day of the month is because of when Crossref Metadata snapshots are released:
> > > "New snapshots are created each month, available by the 5th day, providing all records up to and including the previous month."
> > 
> > 
> > See here: https://www.crossref.org/documentation/metadata-plus/metadata-plus-snapshots/#00344
> 
> I see, I have changed this back :-)
> 

Thanks :)

> > I'm not actually sure how well they keep to that schedule though. Could we add some logging to the check_release task that prints out all of the available releases?
> 
> I'll have a look into adding some logging to the check_release task. I think we will have to get the HEAD of each URL to check if it exists, maybe we could limit it to 2 months before and 2 months after? I don't think there is a single URL with an overview all available releases.

You can find all of the years available with this URL:
`https://api.crossref.org/snapshots/monthly/`

And can query all of the releases available for a each year with this URL, e.g. for 2021: `https://api.crossref.org/snapshots/monthly/2021`

> 
> > I think that there is a bug in the existing code on the develop branch, because the execution date is being used to fetch the the Crossref Metadata release. Should it be execution date - 1 month? E.g. the 2021-05 release is uploaded on 2021-06-05, so if the execution date was 2021-06-01, then it would never fetch the 2021-05 release. I think that is a problem because I set the DAG to run weekly. That wouldn't be a problem for monthly though right? Because the execution date would be for the previous month?
> 
> Yeah I think you are right. This will be solved if we set it to '@monthly' and use the 'prev_execution_date' to create the URL. In that case for the execution_date 2021-06-01, it will be running on 2021-07-01 and it will check for the release of 2021-05.
> Alternatively we could set the schedule to '0 0 7 * *' (so it is after the 5th day) and use the 'execution_date' to create the URL. In that case for the execution_date of 2021-06-07, it will be running on 2021-07-07 and it will check for the release of 2021-06. So we would be a bit more up date.
> 
> > If we set the `schedule_interval` to `monthly`, then that will affect 1) the DOI workflow, which runs weekly and has a sensor that waits for the Crossref Metadata DAG to finish. I think that the sensor will wait indefinitely on weeks when the Crossref Metadata DAG does not run. 2) The worker VM is required,
> > Would you be able to find out how you can get the Crossref Metadata sensor in the DOI DAG to only wait on weeks when the Crossref Metadata DAG runs?
> 
> Yeah by default the ExternalTaskSensor expects the schedule_interval+start date to be in sync. We could change this by using the 'execution_date_fn' parameter of the ExternalTaskSensor. I have added a solution in the latest commit.
> What do you think of this solution? I have tested it with this simple DAG:
> 
> ```python
> from datetime import datetime
> 
> from airflow import DAG
> from airflow.operators.dummy_operator import DummyOperator
> from airflow.sensors.external_task_sensor import ExternalTaskSensor
> from observatory.platform.telescopes.telescope import Telescope, Release
> from observatory.platform.utils.telescope_utils import get_prev_execution_date
> 
> 
> class FakeCrossrefTelescope(Telescope):
>     DAG_ID = 'fake_crossref'
>     SCHEDULE_INTERVAL = '@monthly'
> 
>     def __init__(self, dag_id: str = DAG_ID, start_date: datetime = datetime(2020, 6, 7),
>                  schedule_interval: str = SCHEDULE_INTERVAL, catchup=True):
>         """ Construct a Telescope instance.
> 
>         :param dag_id: the id of the DAG.
>         :param start_date: the start date of the DAG.
>         :param schedule_interval: the schedule interval of the DAG.
>         """
>         super().__init__(dag_id, start_date, schedule_interval, catchup=catchup)
> 
>         self.add_setup_task_chain([self.setup_task])
>         self.add_task(self.task)
> 
>     def make_release(self, **kwargs) -> Release:
>         """ Create a single release instance.
> 
>         :param kwargs: the context passed from the PythonOperator. See
>         https://airflow.apache.org/docs/stable/macros-ref.html for more info.
>         :return: A release instance
>         """
>         release_id = f"telescope_{datetime.now()}"
>         release = Release(self.dag_id, release_id)
>         return release
> 
>     def setup_task(self, **kwargs) -> bool:
>         """ This is a custom task that is executed before a release is made. The return value is used for a
>         shortcircuit operator"
> 
>         :param kwargs: the context passed from the PythonOperator. See
>         https://airflow.apache.org/docs/stable/macros-ref.html for more info.
>         :return: True/False
>         """
>         setup = 'successful'
>         return True if setup == 'successful' else False
> 
>     def task(self, release: Release, **kwargs):
>         """ This is a custom task that can use a release instance.
> 
>         :param release: A release instance.
>         :param kwargs: the context passed from the PythonOperator. See
>         https://airflow.apache.org/docs/stable/macros-ref.html for more info.
>         :return: None.
>         """
>         print(f"This is a custom task that uses the release: {release.release_id}")
> 
> 
> telescope = FakeCrossrefTelescope()
> globals()[telescope.dag_id] = telescope.make_dag()
> 
> 
> default_args = {
>     "owner": "airflow",
>     "start_date": datetime(2020, 8, 30)
> }
> SENSOR_ID_CROSSREF_METADATA = 'crossref_metadata_sensor'
> 
> with DAG(dag_id='test_sensor', schedule_interval='@weekly', default_args=default_args, catchup=False) as dag2:
>     # Sensors
>     crossref_metadata_sensor = ExternalTaskSensor(
>         task_id=SENSOR_ID_CROSSREF_METADATA,
>         external_dag_id=FakeCrossrefTelescope.DAG_ID,
>         execution_date_fn=lambda execution_date: get_prev_execution_date(FakeCrossrefTelescope.SCHEDULE_INTERVAL,
>                                                                          execution_date),
>         mode='reschedule')
> 
>     dummy = DummyOperator(
>         task_id='dummy'
>     )
> 
>     crossref_metadata_sensor >> dummy
> ```
> 
> What would be the issue for the worker VM if the schedule interval is not set to '@Weekly'?

That solution looks good! 

In terms of the worker VM, the only thing is that the Crossref Metadata DAG will not run until the worker VM gets turned on. It will be scheduled to run and wait until it turns on. It would always be the Sunday after the 7th of the month. I suppose that would actually work out OK?

Reply:
Thanks, I somehow did not realise this url was available to check the available releases (https://api.crossref.org/snapshots/monthly/2021). I've added some extra logging on the available releases.

I've changed the schedule interval to '0 0 7 * *', so if the worker VM is turned on weekly it will only run when the snapshot of the previous month is already available.